torchao/prototype/mx_formats/custom_cast.py

```diff
@@ -1080,3 +1082,498 @@ def _(uint8_data):
     def pack_uint6(uint8_data: torch.Tensor) -> torch.Tensor:
         # Dummy placeholder op for torch < 2.4
         raise AssertionError("fp6 packing unsupported without torch >= 2.4")
+
+
+if TORCH_VERSION_AT_LEAST_2_4 and has_triton():
+    import triton
+    import triton.language as tl
+
+    @triton.jit
+    def _triton_calculate_scale(x, axis):
+        # We use a small epsilon to avoid division by zero
+        epsilon = 1e-10
+
+        # TODO(before land): reuse the constants below instead of hardcoding
+        target_max_pow2 = 8
+        e8m0_exponent_bias = 127
+        # bf16_mbits = 7
+        # bf16_exp_bias = 127
+
+        # Find the maximum absolute value for each row
+        max_abs = tl.max(x, axis=axis)
+        # return max_abs, max_abs.to(tl.uint8)
+
+        # TODO 1: rewrite as bit shifts, see https://github.com/pytorch/ao/pull/1908/files
+        # before: 1.7 TB/s
+        # after: ?
+        scale_e8m0_unbiased = tl.floor(tl.log2(max_abs + epsilon)) - target_max_pow2
```

Comment:
I would suspect theres a better way of getting the scale avoiding doing a float log. i guess this is what the TODO is for.